File: native-modules/react-native-get-random-values/android/src/main/java/com/margelo/nitro/reactnativegetrandomvalues/ReactNativeGetRandomValues.kt
package com.margelo.nitro.reactnativegetrandomvalues

import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.nativelogger.OneKeyLog

@DoNotStrip
class ReactNativeGetRandomValues : HybridReactNativeGetRandomValuesSpec() {
  companion object {
    private const val MAX_BYTE_LENGTH = 65536  // 64 KB upper bound
  }

  override fun getRandomBase64(byteLength: Double): String {
    if (byteLength.isNaN() || byteLength.isInfinite() || byteLength != kotlin.math.floor(byteLength)) {
      OneKeyLog.warn("RandomValues", "byteLength must be an integer value, got: $byteLength")
      throw IllegalArgumentException("byteLength must be an integer value")
    }
    val length = byteLength.toInt()
    if (length <= 0 || length > MAX_BYTE_LENGTH) {
      OneKeyLog.warn("RandomValues", "Invalid byteLength: $byteLength, must be 1..$MAX_BYTE_LENGTH")
      throw IllegalArgumentException("Invalid byteLength: must be 1...$MAX_BYTE_LENGTH")
    }
    try {
      val data = ByteArray(length)
      java.security.SecureRandom().nextBytes(data)
      return android.util.Base64.encodeToString(data, android.util.Base64.NO_WRAP)
    } catch (e: Exception) {
      OneKeyLog.error("RandomValues", "SecureRandom failed: ${e.message}\n${e.stackTraceToString()}")
      throw e
    }
  }
}

File: native-modules/react-native-get-random-values/ios/ReactNativeGetRandomValues.swift
import NitroModules
import ReactNativeNativeLogger

class ReactNativeGetRandomValues: HybridReactNativeGetRandomValuesSpec {
    private static let maxByteLength = 65536  // 64 KB upper bound

    func getRandomBase64(byteLength: Double) throws -> String {
        guard byteLength.rounded() == byteLength, !byteLength.isNaN, !byteLength.isInfinite else {
            OneKeyLog.warn("RandomValues", "byteLength must be an integer value, got: \(byteLength)")
            throw NSError(domain: "ReactNativeGetRandomValues", code: -1,
                          userInfo: [NSLocalizedDescriptionKey: "byteLength must be an integer value"])
        }
        let length = Int(byteLength)
        guard length > 0, length <= ReactNativeGetRandomValues.maxByteLength,
              let data = NSMutableData(length: length) else {
            OneKeyLog.warn("RandomValues", "Invalid byteLength: \(byteLength), must be 1...\(ReactNativeGetRandomValues.maxByteLength)")
            throw NSError(domain: "ReactNativeGetRandomValues", code: -1,
                          userInfo: [NSLocalizedDescriptionKey: "Invalid byteLength: must be 1...\(ReactNativeGetRandomValues.maxByteLength)"])
        }
        let result = SecRandomCopyBytes(kSecRandomDefault, length, data.mutableBytes)
        if result != errSecSuccess {
            OneKeyLog.error("RandomValues", "SecRandomCopyBytes failed with status: \(result)")
            throw NSError(domain: "ReactNativeGetRandomValues", code: Int(result), userInfo: nil)
        }
        return data.base64EncodedString(options: [])
    }
}
